rfc/0000-batch-job-groups.rst

```diff
@@ -0,0 +1,1179 @@
+
+
+-------------------
+Costs and Drawbacks
+-------------------
+
+The development cost for this feature is high although substantial
+prototyping has already been done in this space. There are a lot of
+places in the code base this feature touches such as the database
+tables, triggers, and stored procedures, the new REST API interface
+and implementation on the Batch front end, how the driver handles
+cancellation in the scheduler, autoscaler, and canceller, and all of
+the Python and Scala client libraries. Writing tests for this feature
+is time consuming as there are a lot of cases to consider because we
+have a number of different code paths for creating and updating a
+batch and we want to make sure billing and cancellation are done
+properly in different scenarios. In addition, any UI changes are
+extremely time consuming because they cannot be easily tested. The UI
+changes will come **later** on.
+
+Compared to previous features such as open batches, this proposal does
+not require extensive, long running database migrations to transform
+existing tables. The only complicated part is to parameterize the
+existing ``user_inst_coll_resources`` table with the batch ID or
+create a new table entirely by scanning the batches table with an
+explicit lock. It would be easiest to create a separate table entirely.
+
+Other challenges are to make sure the SQL aggregation triggers are
+correctly implemented and the more complicated autoscaler, scheduler,
+and canceller SQL queries are written correctly. However, this would
+be the case for any plan that implemented job groups.
+
+Backwards compatibility is not an issue in this plan.
+
+The way this feature is designed in this proposal will make it easy to
+add components in smaller chunks and the full vision does not need to
+be realized in order to provide QoB with the necessary features it
+needs. However, I am concerned that regardless of how small the actual
+conceptual change is, the number of lines and distinct places this
+change will touch in the code base will make the review process
+challenging. There is tension between breaking up changes into smaller
+chunks and having the entire vision fleshed out and working. We will
+either have to accept larger PRs or accept that there could be bugs
+that are found in later PRs that will need to be fixed that would have
+been caught if we were merging a fully working feature all at once.
+
+The maintenance costs for this feature are moderate. There is another
+level of abstraction in our data model that must be accounted for when
+adding new features in the future or planning a future rewrite of the
+entire system. The UI will also need to be more complicated when we
+expose a nested directory hierarchy to the users.
+
+The proposed simplifications to how cancellation are done will
+increase future developer productivity as this has always been a
+tricky and confusing part of our system especially with how it relates
+to always_run jobs.
```

Comment:
What are the proposed simplifications?

It feels like this RFC actually has two things in it. Maybe we need two RFCs one to address the complexity of cancellation and then a subsequent RFC to address just job groups as an additional change on top of that.